esphome/components/sps30/sps30.cpp

```diff
@@ -0,0 +1,256 @@
+    this->error_code_ = FIRMWARE_VERSION_REQUEST_FAILED;
+    this->mark_failed();
+    return;
+  }
+
+  uint16_t raw_firmware_version[4];
+  if (!this->read_data_(raw_firmware_version, 4)) {
+    this->error_code_ = FIRMWARE_VERSION_READ_FAILED;
+    this->mark_failed();
+    return;
+  }
+  ESP_LOGD(TAG, "  Firmware version v%0d.%02d", (uint16_t(raw_firmware_version[0]) >> 8),
+           uint16_t(raw_firmware_version[0] & 0xFF));
+  /// Serial number identification
+  if (!this->write_command_(SPS30_CMD_GET_SERIAL_NUMBER)) {
+    this->error_code_ = SERIAL_NUMBER_REQUEST_FAILED;
+    this->mark_failed();
+    return;
+  }
+
+  uint16_t raw_serial_number[8];
+  if (!this->read_data_(raw_serial_number, 8)) {
+    this->error_code_ = SERIAL_NUMBER_READ_FAILED;
+    this->mark_failed();
+    return;
+  }
+
+  for (size_t i = 0; i < 8; ++i) {
+    this->serial_number_[i * 2] = uint16_t(uint16_t(raw_serial_number[i]) >> 8);
+    this->serial_number_[i * 2 + 1] = uint16_t(uint16_t(raw_serial_number[i] & 0xFF));
+  }
+  ESP_LOGD(TAG, "  Serial Number: '%s'", this->serial_number_);
+
+  this->start_continuous_measurement_();
+}
+
+void SPS30Component::dump_config() {
+  ESP_LOGCONFIG(TAG, "sps30:");
+  LOG_I2C_DEVICE(this);
+  if (this->is_failed()) {
+    switch (this->error_code_) {
+      case COMMUNICATION_FAILED:
+        ESP_LOGW(TAG, "Communication failed! Is the sensor connected?");
+        break;
+      case MEASUREMENT_INIT_FAILED:
+        ESP_LOGW(TAG, "Measurement Initialization failed!");
+        break;
+      case SERIAL_NUMBER_REQUEST_FAILED:
+        ESP_LOGW(TAG, "Unable to request sensor serial number");
+        break;
+      case SERIAL_NUMBER_READ_FAILED:
+        ESP_LOGW(TAG, "Unable to read sensor serial number");
+        break;
+      case FIRMWARE_VERSION_REQUEST_FAILED:
+        ESP_LOGW(TAG, "Unable to request sensor firmware version");
+        break;
+      case FIRMWARE_VERSION_READ_FAILED:
+        ESP_LOGW(TAG, "Unable to read sensor firmware version");
+        break;
+      default:
+        ESP_LOGW(TAG, "Unknown setup error!");
+        break;
+    }
+  }
+  LOG_UPDATE_INTERVAL(this);
+  ESP_LOGCONFIG(TAG, "  Serial Number: '%s'", this->serial_number_);
+  LOG_SENSOR("  ", "PM1.0", this->pm_1_0_sensor_);
+  LOG_SENSOR("  ", "PM2.5", this->pm_2_5_sensor_);
+  LOG_SENSOR("  ", "PM4", this->pm_4_0_sensor_);
+  LOG_SENSOR("  ", "PM10", this->pm_10_0_sensor_);
+}
+
+void SPS30Component::update() {
+  /// Check if warning flag active (sensor reconnected?)
+  if (this->status_has_warning()) {
+    ESP_LOGD(TAG, "Retrying to reconnect the sensor.");
+    if (this->write_command_(SPS30_CMD_SOFT_RESET)) {
+      /// Sensor restarted and reading attempt made next cycle
+      delayMicroseconds(500000);
+      this->start_continuous_measurement_();
+      this->status_clear_warning();
+      ESP_LOGD(TAG, "Sensor reconnected!");
+    } else {
+      ESP_LOGD(TAG, "Sensor offline?");
+    }
+  } else {
```

Comment:
```suggestion
    return;
  }
```

Add the return block there so we can save on indentation for the main block of this function.

Reply:
Thanks. I've refactored it.